rclcpp/test/rclcpp/test_qos_event.cpp

```diff
@@ -222,3 +229,97 @@ TEST_F(TestQosEvent, test_default_incompatible_qos_callbacks)
 
   rcutils_logging_set_output_handler(original_output_handler);
 }
+
+TEST_F(TestQosEvent, construct_destruct_rcl_error) {
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>(topic_name, 10);
+  auto rcl_handle = publisher->get_publisher_handle();
+  ASSERT_NE(nullptr, rcl_handle);
+
+  // This callback requires some type of parameter, but it could be anything
+  auto callback = [](int) {};
+  const rcl_publisher_event_type_t event_type = RCL_PUBLISHER_OFFERED_DEADLINE_MISSED;
+
+  {
+    // Logs error and returns
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_publisher_event_init, RCL_RET_ERROR);
+
+    auto throwing_statement = [callback, rcl_handle, event_type]() {
+        // reset() is not needed for the exception, but it handles unused return value warning
+        std::make_shared<
+          rclcpp::QOSEventHandler<decltype(callback), std::shared_ptr<rcl_publisher_t>>>(
```

Comment:
@clalancette Here's a situation where using `auto` shot me in the foot. For some reason with windows, capturing rcl_handle by reference in the lambda and then calling decltype(rcl_handle) caused it to think that this should have been a `shared_ptr<rcl_publisher_t>&`, which then caused it to complain about an uninitialized reference of all things.